Bug 1472226 - Add additional field validations for JSON Schema.

cfchase opened the pull request:
Updated apb spec to allow for more validations from JSON Schema. 
https://datatracker.ietf.org/doc/draft-handrews-json-schema-validation/

Validations follow JSON schema keywords, and uses snake_case instead of camelCase for consistency when writing APBs

Added apb_field (JSON schema field):
- max_length (maxLength)
- min_length (minLength)
- multiple_of (multipleOf)
- minimum (minimum)
- maximum (maximum)
- exclusive_minimum (exclusiveMinimum)
- exclusive_maximum (exclusiveMaximum)

Allow maxlength to be used for backwards compatibility as DeprecatedMaxlength.
Allowed `type: integer` since it's used in JSON schema

cfchase commented on pkg/broker/util.go:
```diff
@@ -290,33 +289,87 @@ func extractProperties(params []apb.ParameterDescriptor) (map[string]*schema.Sch
+	if min, ok := pd.Minimum.(float64); ok {
+		prop.Minimum = schema.Number{Val: min, Initialized: true}
+	}
+
```
In the JSON schema [validations documentation](https://datatracker.ietf.org/doc/draft-handrews-json-schema-validation/), `exclusiveMinimum` and `exclusiveMaximum` MUST be a number.  However go-jsschema and the serialized values it sends uses boolean values for `exclusiveMaximum` and `exclusiveMinimum` are boolean and use the numbers in `maximum` and `minimum`.  

I followed with the docs and  used `exclusive_maximum`/`exclusive_minimum` as a number for the APB spec, but if there are strong opinions that I should use boolean `exclusive_maximum`/`exclusive_minimum` instead, I'd like to hear reasoning.  I was divided myself.

dymurray replied:
I agree with following the docs. `exclusive_maximum` representing the max number seems intuitive to me.

shawn-hurley commented on pkg/apb/types.go:
```diff
@@ -34,18 +34,30 @@ type SpecManifest map[string]*Spec
+
+	// number validators
+	MultipleOf       float64     `json:"multipleOf,omitempty" yaml:"multiple_of,omitempty"`
+	Maximum          interface{} `json:"maximum,omitempty"`
```
I don't like the loss of type validation if someone is creating a ParameterDescription by hand. If we can make this not exported I would be much less worried about it. 

This is another way around this. Using a type declaration for float64 and dereferencing the pointer when using that value. Example: 

```go
type nilableNumber float64

type ParameterDescription struct {
.....
	Maximum          *nilableNumber `json:"maximum,omitempty"`
.....
}
```

https://play.golang.org/p/qpeFB6b8GXp

cfchase commented:
@shawn-hurley thanks for the tip.  I switched to the nilableNumber in https://github.com/openshift/ansible-service-broker/pull/615/commits/7450587c1271c57fd9b1f45e45f33f3981d7cf92.

Let me know if this addresses your concerns.

shawn-hurley commented on pkg/broker/util.go:
```diff
@@ -290,33 +289,87 @@ func extractProperties(params []apb.ParameterDescriptor) (map[string]*schema.Sch
+	}
+
+	// max_length overrides maxlength
+	if pd.MaxLength > 0 {
```
I am pretty sure that you are going to need to dereference the pointer here.

shawn-hurley commented on pkg/apb/types.go:
```diff
@@ -32,20 +32,34 @@ type Parameters map[string]interface{}
 // SpecManifest - Spec ID to Spec manifest
 type SpecManifest map[string]*Spec
 
+type nilableNumber float64
```
I think you should make this exportable so that it can be used when creating a ParameterDescriptor outside of the APB package.
